File: CELEBAL/linkedlist.java
class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    private Node head;

    public LinkedList() {
        head = null;
    }

    // Append a new node at the end
    public void append(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            return;
        }

        Node current = head;
        while (current.next != null) {
            current = current.next;
        }

        current.next = newNode;
    }

    // Print the entire list
    public void printList() {
        if (head == null) {
            System.out.println("LinkedList is empty.");
            return;
        }

        Node current = head;
        while (current != null) {
            System.out.print(current.data);
            if (current.next != null)
                System.out.print(" → ");
            current = current.next;
        }
        System.out.println();
    }

    // Delete the nth node (1-based index)
    public void deleteNthNode(int n) throws Exception {
        if (head == null) {
            throw new Exception("Cannot delete from an empty list.");
        }

        if (n <= 0) {
            throw new Exception("Invalid index " + n + ". Index must be >= 1.");
        }

        if (n == 1) {
            head = head.next;
            return;
        }

        Node prev = head;
        int count = 1;

        while (prev != null && count < n - 1) {
            prev = prev.next;
            count++;
        }

        if (prev == null || prev.next == null) {
            throw new Exception("Index " + n + " is out of range.");
        }

        prev.next = prev.next.next;
    }
}

public class LinkedListDemo {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        // Add nodes
        list.append(10);
        list.append(20);
        list.append(30);
        list.append(40);
        list.append(50);

        System.out.println("Initial list:");
        list.printList(); // 10 → 20 → 30 → 40 → 50

        // Delete 1st node
        try {
            list.deleteNthNode(1);
            System.out.println("\nAfter deleting 1st node:");
            list.printList(); // 20 → 30 → 40 → 50
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        // Delete 3rd node
        try {
            list.deleteNthNode(3);
            System.out.println("\nAfter deleting 3rd node:");
            list.printList(); // 20 → 30 → 50
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        // Delete last node
        try {
            list.deleteNthNode(3);
            System.out.println("\nAfter deleting last node:");
            list.printList(); // 20 → 30
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        // Out-of-range deletion
        try {
            list.deleteNthNode(5);
        } catch (Exception e) {
            System.out.println("\nError on deleteNthNode(5): " + e.getMessage());
        }

        // Delete remaining nodes
        try {
            list.deleteNthNode(1); // delete 20
            list.deleteNthNode(1); // delete 30
            System.out.println("\nAfter deleting all nodes:");
            list.printList(); // LinkedList is empty
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        // Try deleting from empty list
        try {
            list.deleteNthNode(1);
        } catch (Exception e) {
            System.out.println("\nError on deleteNthNode(1) when empty: " + e.getMessage());
        }
    }
}

File: CELEBAL/linkedlist.py
class Node:
    """
    A Node in a singly linked list.
    Holds 'data' and a reference to the 'next' node.
    """
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    """
    A singly linked list implementation.
    Supports:
      - append(data): add a new node with 'data' at the end
      - print_list(): print all node values in sequence
      - delete_nth_node(n): delete the node at 1-based index n
    """
    def __init__(self):
        self.head = None

    def append(self, data):
        """
        Append a new node with the given data to the end of the list.
        """
        new_node = Node(data)
        if not self.head:
            # Empty list → new_node becomes head
            self.head = new_node
            return

        # Traverse to the end
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def print_list(self):
        """
        Traverse the list and print each node's data.
        """
        if not self.head:
            print("LinkedList is empty.")
            return

        current = self.head
        nodes = []
        while current:
            nodes.append(str(current.data))
            current = current.next
        print(" → ".join(nodes))

    def delete_nth_node(self, n):
        """
        Delete the node at 1-based index n.
        Raises:
          - IndexError if the list is empty or n is out of valid range.
        """
        # Edge case: deleting from an empty list
        if not self.head:
            raise IndexError("Cannot delete from an empty list.")

        # Edge case: non‐positive index
        if n <= 0:
            raise IndexError(f"Invalid index {n}. Index must be >= 1.")

        # If deleting the head (n == 1)
        if n == 1:
            to_delete = self.head
            self.head = self.head.next
            to_delete.next = None  # Help the garbage collector
            return

        # Traverse to the (n-1)-th node, so we can skip its 'next'
        prev = self.head
        current_index = 1
        while prev and current_index < n - 1:
            prev = prev.next
            current_index += 1

        # If prev is None or prev.next is None, n is out of range
        if prev is None or prev.next is None:
            raise IndexError(f"Index {n} is out of range.")

        # Delete the n-th node
        to_delete = prev.next
        prev.next = to_delete.next
        to_delete.next = None  # Help the garbage collector


if __name__ == "__main__":
    # ----------------------------
    # Example usage / Test cases
    # ----------------------------

    ll = LinkedList()

    # 1) Append some items
    for value in [10, 20, 30, 40, 50]:
        ll.append(value)

    print("Initial list:")
    ll.print_list()          # Expected: 10 → 20 → 30 → 40 → 50

    # 2) Delete the 1st node (head)
    try:
        ll.delete_nth_node(1)
        print("\nAfter deleting 1st node:")
        ll.print_list()      # Expected: 20 → 30 → 40 → 50
    except IndexError as e:
        print(f"Error: {e}")

    # 3) Delete the 3rd node (which is currently '40')
    try:
        ll.delete_nth_node(3)
        print("\nAfter deleting 3rd node:")
        ll.print_list()      # Expected: 20 → 30 → 50
    except IndexError as e:
        print(f"Error: {e}")

    # 4) Delete the last node (now index = length = 3)
    try:
        ll.delete_nth_node(3)
        print("\nAfter deleting last node:")
        ll.print_list()      # Expected: 20 → 30
    except IndexError as e:
        print(f"Error: {e}")

    # 5) Attempt to delete with an out-of-range index (e.g., 5)
    try:
        ll.delete_nth_node(5)
    except IndexError as e:
        print(f"\nError on delete_nth_node(5): {e}")

    # 6) Delete remaining nodes one by one
    try:
        ll.delete_nth_node(1)   # deletes 20
        ll.delete_nth_node(1)   # deletes 30, now list is empty
        print("\nAfter deleting all nodes:")
        ll.print_list()         # Expected: "LinkedList is empty."
    except IndexError as e:
        print(f"Error: {e}")

    # 7) Attempt to delete from an empty list
    try:
        ll.delete_nth_node(1)
    except IndexError as e:
        print(f"\nError on delete_nth_node(1) when empty: {e}")
